File: core/license.py
"""
core/license.py
===============
The node-lock: bind an installation to one physical machine.

A `license.key` file contains:
  * the HARDWARE ID this install is licensed to,
  * issue metadata,
  * an HMAC-SHA256 signature over the payload keyed by SECRET_SALT.

On every boot the core:
  1. recomputes the live HARDWARE ID from the host (core/hardware.py),
  2. loads license.key,
  3. verifies the signature (proves the file was produced by something holding
     SECRET_SALT -- i.e. the genuine vendor tooling / compiled core),
  4. verifies the licensed HARDWARE ID == the live HARDWARE ID.

Copy the whole install folder to a different laptop and step 4 fails: the
license is cryptographically welded to the original silicon.
"""

from __future__ import annotations

import hashlib
import hmac
import json
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path

from core import config
from core.hardware import resolve_identity


class LicenseError(Exception):
    """Raised when the license is missing, malformed, forged, or node-mismatched."""


@dataclass(frozen=True)
class LicenseStatus:
    activated: bool
    hardware_id: str
    reason: str = ""


def _sign_payload(payload: dict) -> str:
    """Deterministic HMAC-SHA256 over the canonical JSON payload."""
    canonical = json.dumps(payload, sort_keys=True, separators=(",", ":")).encode()
    return hmac.new(config.SECRET_SALT, canonical, hashlib.sha256).hexdigest()


def generate_license(target_hardware_id: str | None = None) -> dict:
    """Create and persist a license.key bound to the given (or live) hardware id.

    Called by setup.sh / run.py so the first boot is already unlocked on the
    builder's own machine.
    """
    config.ensure_state_dir()

    hardware_id = target_hardware_id or resolve_identity().hardware_id

    payload = {
        "product": config.PRODUCT_NAME,
        "edition": config.PRODUCT_EDITION,
        "version": config.PRODUCT_VERSION,
        "hardware_id": hardware_id,
        "issued_at": datetime.now(timezone.utc).isoformat(),
    }
    document = {"payload": payload, "signature": _sign_payload(payload)}

    config.LICENSE_FILE.write_text(json.dumps(document, indent=2), encoding="utf-8")
    return document


def _load_document(path: Path) -> dict:
    try:
        return json.loads(path.read_text(encoding="utf-8"))
    except FileNotFoundError as exc:
        raise LicenseError("No license.key present -- product is locked.") from exc
    except (json.JSONDecodeError, OSError) as exc:
        raise LicenseError("license.key is unreadable or corrupt.") from exc


def verify_license() -> LicenseStatus:
    """Verify the on-disk license against the live hardware. Never raises."""
    live_id = resolve_identity().hardware_id

    try:
        document = _load_document(config.LICENSE_FILE)
        payload = document["payload"]
        signature = document["signature"]
    except LicenseError as exc:
        return LicenseStatus(activated=False, hardware_id=live_id, reason=str(exc))
    except (KeyError, TypeError):
        return LicenseStatus(
            activated=False, hardware_id=live_id, reason="license.key is malformed."
        )

    # 1. Signature integrity -- constant-time compare defeats forgery attempts.
    expected = _sign_payload(payload)
    if not hmac.compare_digest(expected, str(signature)):
        return LicenseStatus(
            activated=False,
            hardware_id=live_id,
            reason="License signature invalid -- file was tampered with or forged.",
        )

    # 2. Node-lock: licensed hardware must equal live hardware.
    licensed_id = str(payload.get("hardware_id", ""))
    if not hmac.compare_digest(licensed_id, live_id):
        return LicenseStatus(
            activated=False,
            hardware_id=live_id,
            reason="Hardware mismatch -- this license is bound to a different machine.",
        )

    return LicenseStatus(activated=True, hardware_id=live_id, reason="OK")


def require_activation() -> str:
    """Return the live hardware id if activated, else raise LicenseError."""
    status = verify_license()
    if not status.activated:
        raise LicenseError(status.reason)
    return status.hardware_id


if __name__ == "__main__":  # pragma: no cover
    st = verify_license()
    print("Activated:", st.activated)
    print("Hardware :", st.hardware_id)
    print("Reason   :", st.reason)

File: core-rs/src/license.rs
//! Node-locked license (Rust port of core/license.py).
//!
//! A `license.key` binds the install to one machine's HARDWARE ID and is signed
//! with HMAC-SHA256 keyed by the secret salt. On boot the Rust core recomputes
//! the live HARDWARE ID and verifies both the signature and the node match,
//! using a constant-time comparison.

use std::path::Path;

use hmac::{Hmac, Mac};
use serde::{Deserialize, Serialize};
use sha2::Sha256;

use crate::config;

type HmacSha256 = Hmac<Sha256>;

#[derive(Debug, Serialize, Deserialize)]
struct LicensePayload {
    product: String,
    edition: String,
    version: String,
    hardware_id: String,
    issued_at: String,
}

#[derive(Debug, Serialize, Deserialize)]
struct LicenseDocument {
    payload: LicensePayload,
    signature: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct LicenseStatus {
    pub activated: bool,
    pub hardware_id: String,
    pub reason: String,
}

fn sign_payload(payload: &LicensePayload, salt: &[u8]) -> String {
    // serde_json serializes struct fields in declaration order -> deterministic,
    // so re-serializing a parsed payload yields identical bytes for verification.
    let canonical = serde_json::to_vec(payload).expect("serialize license payload");
    let mut mac = HmacSha256::new_from_slice(salt).expect("hmac key");
    mac.update(&canonical);
    hex::encode(mac.finalize().into_bytes())
}

/// Constant-time string comparison.
fn ct_eq(a: &str, b: &str) -> bool {
    let (a, b) = (a.as_bytes(), b.as_bytes());
    if a.len() != b.len() {
        return false;
    }
    let mut diff: u8 = 0;
    for (x, y) in a.iter().zip(b.iter()) {
        diff |= x ^ y;
    }
    diff == 0
}

fn license_path(state_dir: &Path) -> std::path::PathBuf {
    state_dir.join("license.key")
}

/// Generate and persist a license bound to `hardware_id`.
pub fn generate(state_dir: &Path, salt: &[u8], hardware_id: &str) -> std::io::Result<()> {
    std::fs::create_dir_all(state_dir)?;
    let payload = LicensePayload {
        product: config::PRODUCT_NAME.to_string(),
        edition: config::PRODUCT_EDITION.to_string(),
        version: config::PRODUCT_VERSION.to_string(),
        hardware_id: hardware_id.to_string(),
        issued_at: chrono::Utc::now().to_rfc3339(),
    };
    let signature = sign_payload(&payload, salt);
    let document = LicenseDocument { payload, signature };
    let json = serde_json::to_string_pretty(&document).expect("serialize license");
    std::fs::write(license_path(state_dir), json)
}

/// Verify the on-disk license against `live_hardware_id`. Never panics.
pub fn verify(state_dir: &Path, salt: &[u8], live_hardware_id: &str) -> LicenseStatus {
    let locked = |reason: &str| LicenseStatus {
        activated: false,
        hardware_id: live_hardware_id.to_string(),
        reason: reason.to_string(),
    };

    let raw = match std::fs::read_to_string(license_path(state_dir)) {
        Ok(r) => r,
        Err(_) => return locked("No license.key present — product is locked."),
    };
    let document: LicenseDocument = match serde_json::from_str(&raw) {
        Ok(d) => d,
        Err(_) => return locked("license.key is unreadable or corrupt."),
    };

    let expected = sign_payload(&document.payload, salt);
    if !ct_eq(&expected, &document.signature) {
        return locked("License signature invalid — file was tampered with or forged.");
    }
    if !ct_eq(&document.payload.hardware_id, live_hardware_id) {
        return locked("Hardware mismatch — this license is bound to a different machine.");
    }

    LicenseStatus {
        activated: true,
        hardware_id: live_hardware_id.to_string(),
        reason: "OK".to_string(),
    }
}
